### ISSUE TITLE
total_ordering of a class wrongly impacts the type restriction behavior

### ISSUE THREAD

**Issue by Questionsdenombres** [microsoft/pylance-release#3033]
## Environment data
- Pylance v2022.7.20
- Python 3.10.5 (tags/v3.10.5:f377153, Jun  6 2022, 16:14:13) - python.org distribution 
- Visual studio: 1.69.0
- Windows-10-10.0.22000-SP0

## Code Snippet
```python
from functools import total_ordering

@total_ordering
class X:
    def __eq__(self, other: object) -> bool: ...
    def __lt__(self, other: object) -> bool: ...

def X_func(x: X | tuple[bool, int]) -> None:
    if isinstance(x, X):
        print(x)
    else:
        (x1, x2) = x
        print((x1, x2))
    return
```
## Expected behavior
-  no typechecker report
-  variable x1: bool
-  variable x2: int

## Actual behavior
The typechecker reports:
- Type of "x1" is partially unknown (variable x1: Unknown | bool)
- Type of "x2" is partially unknown (variable x2: Unknown | int)
- and consequently: "X" is not iterable.

**Important**: the reports disappear when commenting the 'total_ordering' line.

## Thanks ...
... for all the great job you're doing.
I can't imagine python programming without a type checker.
And Pylance is a very good one.
Patrick.

**Comment by erictraut** [comment]
Thanks for the bug report. Someone else reported this issue last week in the [pyright issue tracker](https://github.com/microsoft/pyright/issues/3633), and it has already been addressed in pyright. The fix will appear in the next release of pylance.

**Comment by heejaechang** [comment]
This issue has been fixed in version 2022.7.30, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2022730-13-July-2022)
